Fix nodes volumesAttached status not being updated

jfremy opened the pull request:
/kind bug

Fixes #107973 

The UpdateNodeStatuses code stops too early in case there is an error when calling updateNodeStatus. It will return immediately which means any remaining node won't have its update status put back to true.

Looking at the call sites for UpdateNodeStatuses, it appears this is not the only issue. If the lister call fails with anything but a Not Found error, it's silently ignored which is wrong in the detach path.
Also the reconciler detach path calls UpdateNodeStatuses but the real intent is to only update the node currently processed in the loop and not proceed with the detach call if there is an error updating that specify node volumesAttached property. With the current implementation, it will not proceed if there is an error updating another node (which is not completely bad but not ideal) and worse it will proceed if there is a lister error on that node which means the node volumesAttached property won't have been updated.

To fix those issues, introduce the following changes:
- [node_status_updater] introduce UpdateNodeStatusForNode which does what UpdateNodeStatuses does but only for the provided node
- [node_status_updater] if the node lister call fails for anything but a Not Found error, we will return an error, not ignore it
- [node_status_updater] if the update of a node volumesAttached properties fails we continue processing the other nodes
- [actual_state_of_world] introduce GetVolumesToReportAttachedForNode which does what GetVolumesToReportAttached but for the node whose name is provided it returns a bool which indicates if the node in question needs an update as well as the volumesAttached list. It is used by UpdateNodeStatusForNode
- [reconciler] use UpdateNodeStatusForNode in the detach loop

```release-note
Fix a bug in attachdetach controller that didn't properly handle kube-apiserver errors leading to stuck attachments/detachments.
```

jfremy committed:
Fix nodes volumesAttached status not updated

The UpdateNodeStatuses code stops too early in case there is
an error when calling updateNodeStatus. It will return immediately
which means any remaining node won't have its update status put back
to true.

Looking at the call sites for UpdateNodeStatuses, it appears this is
not the only issue. If the lister call fails with anything but a Not Found
error, it's silently ignored which is wrong in the detach path.
Also the reconciler detach path calls UpdateNodeStatuses but the real intent
is to only update the node currently processed in the loop and not proceed
with the detach call if there is an error updating that specifi node volumesAttached
property. With the current implementation, it will not proceed if there is
an error updating another node (which is not completely bad but not ideal) and
worse it will proceed if there is a lister error on that node which means the
node volumesAttached property won't have been updated.

To fix those issues, introduce the following changes:
- [node_status_updater] introduce UpdateNodeStatusForNode which does what
  UpdateNodeStatuses does but only for the provided node
- [node_status_updater] if the node lister call fails for anything but a Not
  Found error, we will return an error, not ignore it
- [node_status_updater] if the update of a node volumesAttached properties fails
  we continue processing the other nodes
- [actual_state_of_world] introduce GetVolumesToReportAttachedForNode which
  does what GetVolumesToReportAttached but for the node whose name is provided
  it returns a bool which indicates if the node in question needs an update as
  well as the volumesAttached list. It is used by UpdateNodeStatusForNode
- [actual_state_of_world] use write lock in updateNodeStatusUpdateNeeded, we're
  modifying the map content
- [reconciler] use UpdateNodeStatusForNode in the detach loop

jfremy commented:
/retitle Fix nodes volumesAttached status not being updated

gnufied commented:
/ok-to-test

gnufied commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -677,10 +671,49 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
+	// When GetVolumesToReportAttached is called by node status updater, the current status
+	// of this node will be updated, so set the flag statusUpdateNeeded to false indicating
+	// the current status is already updated.
+	if err := asw.updateNodeStatusUpdateNeeded(nodeName, false); err != nil {
```
Looks like we are updating ASW with read lock?

jfremy replied:
Good catch - do you want me to fix this as well?

gnufied commented on pkg/controller/volume/attachdetach/statusupdater/node_status_updater.go:
```diff
@@ -57,40 +60,65 @@ type nodeStatusUpdater struct {
 		}
+	}
+	if len(nodeIssues) > 0 {
+		return fmt.Errorf("%v could not be updated", nodeIssues)
```
We can also return a generic error so as we are not logging 100-300 concatenated node names in logs at once. Something like - `fmt.Errorf("Unable to update %d  nodes for volume attach status")` ? We are anyways logging the error  from individual nodes in `processNodeVolumes` functions.

jfremy replied:
Sure I can update the messages

gnufied commented:
Thanks for the PR. Some unit tests would be nice which can exercise the bug in a reasonable fashion.

jfremy commented:
> Thanks for the PR. Some unit tests would be nice which can exercise the bug in a reasonable fashion.

yes will add

jfremy commented:
> Thanks for the PR. Some unit tests would be nice which can exercise the bug in a reasonable fashion.

Still working on it. There were no unit tests for NodeStatusUpdater so it's a little more work than I expected

wojtek-t commented:
/cc

gnufied commented:
One of the ways you can possibly test drive this is by using something like a fakeClient with a custom reactor for stuff. Something like - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/testing/testing.go#L554 

There should be other similar examples in the code.

jfremy commented:
> One of the ways you can possibly test drive this is by using something like a fakeClient with a custom reactor for stuff. Something like - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/testing/testing.go#L554 
> 
> 
> 
> There should be other similar examples in the code. 

Yes, that's what I've been using. Initially I was looking at the existing fake client factory used in the other unit tests but it did not really fit the need.
I'll have some unit tests for node status updater soon

jfremy committed:
Add unit tests

- actual_state_of_world_test.go: test the new method GetVolumesToReportAttachedForNode
  for an existing node and a non-existing node
- node_status_updater_test.go: test UpdateNodeStatuses and UpdateNodeStatuses in nominal
  case with 2 nodes getting one volume each. Test UpdateNodeStatuses with the first call
  to node.patch failing but the following one succeeding
- add comment in node_status_updater.go
- fix log line in reconciler.go
- rename variable in actual_state_of_world.go

jfremy commented:
> One of the ways you can possibly test drive this is by using something like a fakeClient with a custom reactor for stuff. Something like - https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/persistentvolume/testing/testing.go#L554
> 
> There should be other similar examples in the code.

I've added a few unit tests to check the nominal case and one test that check non-regression on this issue

jfremy commented:
Here's a summary of the fixes / changes included in this PR:
- fix use of read lock in actual_state_of_world.go GetVolumesToReportAttached - it should be write lock since we modify the map content
- fix bug in node_status_updater.go UpdateNodeStatuses that would return an error immediately if there was an error in node.patch
- fix bug in node_status_updater.go UpdateNodeStatuses that would not return an error and skip the node if the NodeLister returned a non IsNotFound error for the node
- changed NodeStatusUpdater to allow processing a single node - this is used in the detach loop of reconciler.go (which process volumes one by one as opposed to the attach section that does every node/volume in one go) - this required adding an matching method in ActualStateOfWorld

YuikoTakada commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -677,10 +671,49 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
+		klog.Errorf("Failed to update statusUpdateNeeded field when getting volumes: %v", err)
+	}
+
+	return true, attachedVolumes
```
ditto

YuikoTakada commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -677,10 +671,49 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
+	asw.Lock()
+	defer asw.Unlock()
+
+	var attachedVolumes []v1.AttachedVolume
```
nit: `volumesToReportAttached` seems to be easy to understand same as https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/cache/actual_state_of_world.go#L653
```suggestion
	var volumesToReportAttached []v1.AttachedVolume
```

YuikoTakada commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -677,10 +671,49 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
+		return false, nil
+	}
+
+	attachedVolumes = asw.getAttachedVolumeFromUpdateObject(nodeToUpdateObj.volumesToReportAsAttached)
```
ditto

YuikoTakada reviewed (commented):
Thank you for your PR. I prefer this approach. I wrote some comments, so please check them.

gnufied commented:
/retest

jfremy commented:
Can a trusted user add a retitle command? 
`/retitle Fix nodes volumesAttached status not being updated`

My initial title did not please the bot. I changed it in github but that does not seem to register and I don't have the rights to do the retitle myself

gnufied commented:
Bot is complaining about commit message in one of your first commits, not the PR title. Can you amend your commit and force push to this branch?

jfremy commented:
> Bot is complaining about commit message in one of your first commits, not the PR title. Can you amend your commit and force push to this branch?

done. thanks! I find the message from the bot confusing

gnufied commented:
Thanks for working on it. Can you please squash commits which are not not directly on topic. Generally - commits such as "PR review comments" etc should be squashed before a PR can merge, so as commit history of change looks reasonable.

jingxu97 commented on pkg/controller/volume/attachdetach/cache/actual_state_of_world.go:
```diff
@@ -677,10 +671,48 @@ func (asw *actualStateOfWorld) GetVolumesToReportAttached() map[types.NodeName][
 	return volumesToReportAttached
 }
 
+func (asw *actualStateOfWorld) GetVolumesToReportAttachedForNode(nodeName types.NodeName) (bool, []v1.AttachedVolume) {
```
it seems to me the returned bool is not very necessary. Based on reconciler logic, the only possible case is the node is no longer in actual state. In this case, volume list will be empty. I am wondering in this case, we should update node status to have empty volume list or not. Or Maybe this case of node empty is also should not exist?

jfremy replied:
If I don't report this bool, then I have no way of knowing if you need to update the node status.volumesAttached property with an empty list or if there is nothing to do (for instance when we detach the last volume from the node)

jingxu97 replied:
What I mean is GetVolumesToReportAttachedForNode() is only needed by UpdateNodeStatusForNode(), which is only called in reconciler during detach, and before calling it, it mark volume/node by RemoveVolumeFromReportAsAttached. In this case, the bool will be always true, right? If this function is used in other places, we need to have bool value, I think.
 https://github.com/kubernetes/kubernetes/pull/108167/files#diff-e76cff1d139f1df6d8c37cea61b19aa8caea6b103d64e58a084f2d81c5bad149R196

jfremy commented:
/retest

gnufied commented:
/lgtm

gnufied commented:
/approve

jfremy commented:
/retest

jingxu97 commented:
also created an issue to follow up to see whether it is better to simply some of the logic here to avoid future mistakes. https://github.com/kubernetes/kubernetes/issues/108433

@jfremy @gnufied what do you think?

mattcary commented:
FYI I'll be creating cherrypicks, reach out to me on slack if questions/comments/concerns.

gnufied commented:
@jingxu97 yeah I think simplifying fetching the node status and update logic makes sense.  Lets do this in a follow up.

ialidzhikov commented:
> FYI I'll be creating cherrypicks, reach out to me on slack if questions/comments/concerns.

Is this fix applicable for 1.21 as well? I see that currently it is only backported to 1.22 and 1.23.

cc @wojtek-t

mattcary commented:
I think it would be, but I forget the context. It applies cleanly to 1.21 at least.

\cc @jingxu97 

Do  you remember, Jing?

wojtek-t commented:
> Is this fix applicable for 1.21 as well? I see that currently it is only backported to 1.22 and 1.23.

Yes - we should backport to 1.21 too.

> I think it would be, but I forget the context. It applies cleanly to 1.21 at least.

Volume attachments/detachments were stuck in case of errors returned from apiserver.
@mattcary - can you please open 1.21 cherrypick since you were opening 1.22 and 1.23 ones?

mattcary commented:
👍🏾

https://github.com/kubernetes/kubernetes/pull/109206
